File: eshop/backend_django/api/serializers/logs.py
from rest_framework import serializers

from api.models import AdminActivityLog, SystemRequestLog, UserActivityLog


class UserActivityLogSerializer(serializers.ModelSerializer):
    product_name = serializers.CharField(source="product.name", read_only=True)
    trader_name = serializers.CharField(source="trader.business_name", read_only=True)

    class Meta:
        model = UserActivityLog
        fields = "__all__"
        read_only_fields = ("user", "username_snapshot", "ip_address", "user_agent", "device_type", "browser", "os", "created_at")

    def validate(self, attrs):
        if attrs.get("action") == UserActivityLog.Action.RATING and (attrs.get("rating_value") is None or not 1 <= attrs["rating_value"] <= 5):
            raise serializers.ValidationError({"rating_value": "Rating must be between 1 and 5."})
        return attrs


class AdminActivityLogSerializer(serializers.ModelSerializer):
    class Meta:
        model = AdminActivityLog
        fields = "__all__"


class SystemRequestLogSerializer(serializers.ModelSerializer):
    class Meta:
        model = SystemRequestLog
        fields = "__all__"

File: eshop/admin_web_nextjs/nextjs-admin/src/types/logs.ts
export interface UserActivityLog {
  id: number; username_snapshot: string; session_key: string; ip_address: string | null; device_type: string;
  browser: string; os: string; action: string; product: number | null; product_name?: string | null;
  rating_value: number | null; search_query: string; metadata: Record<string, unknown>; created_at: string;
}

export interface AdminActivityLog {
  id: number; actor_username: string; ip_address: string | null; device_type: string; browser: string; os: string;
  module: string; action: string; object_type: string; object_id: string; object_repr: string;
  status_code: number | null; metadata: Record<string, unknown>; created_at: string;
}

export interface SystemRequestLog {
  id: number; username_snapshot: string; ip_address: string | null; device_type: string; browser: string; os: string;
  method: string; path: string; status_code: number | null; duration_ms: number | null; is_error: boolean;
  error_message: string; metadata: Record<string, unknown>; created_at: string;
}
